BlueprintUICommonControls/Sources/ScrollView.swift

```diff
@@ -260,8 +300,137 @@ fileprivate final class ScrollerWrapperView: UIView {
                 self.scrollView.contentOffset.x = -contentInset.left
             }
         }
+    }
+    
+    //
+    // MARK: UIView
+    //
+    
+    public override func didMoveToWindow() {
+        super.didMoveToWindow()
+        
+        if self.window != nil {
+            self.updateBottomContentInsetWithKeyboardFrame()
+        }
+    }
+    
+    public override func didMoveToSuperview() {
+        super.didMoveToSuperview()
+        
+        if self.superview != nil {
+            self.updateBottomContentInsetWithKeyboardFrame()
+        }
+    }
+}
 
 
+extension ScrollView
+{
+    // Calculates the correct content inset to apply for the given inputs.
+    
+    static func finalContentInset(
+        scrollViewInsets : UIEdgeInsets,
+        safeAreaInsets : UIEdgeInsets,
+        keyboardBottomInset : CGFloat,
+        refreshControlState : PullToRefreshBehavior,
+        refreshControlBounds : CGRect?
+    ) -> UIEdgeInsets
+    {
+        var finalContentInset = scrollViewInsets
+        
+        // Include the keyboard's adjustment at the bottom of the scroll view.
+        
+        if keyboardBottomInset > 0.0 {
+            finalContentInset.bottom += keyboardBottomInset
+            
+            // Exclude the safe area insets, so the content hugs the top of the keyboard.
+            
+            finalContentInset.bottom -= safeAreaInsets.bottom
+        }
+        
+        // The refresh control lives above the content and adjusts the
+        // content inset for itself when visible and refreshing.
+        // Do the same adjustment to our expected content inset.
+        
+        if case .refreshing = refreshControlState {
+            finalContentInset.top += refreshControlBounds?.size.height ?? 0.0
+        }
+        
+        return finalContentInset
     }
+}
 
+
+extension ScrollerWrapperView : KeyboardObserverDelegate {
+    
+    //
+    // MARK: Keyboard
+    //
+    
+    private func updateBottomContentInsetWithKeyboardFrame() {
```

Comment:
This is almost the same func as `applyContentInset`, can they merge?

Reply:
Doh, will do

Reply:
Need to dig into this more... there's a bit of nuanced behaviour between the logic in `applyContentInset` handling changes only from the model updates, vs `updateBottomContentInsetWithKeyboardFrame` always updating on live views.

Reply:
Going to leave as is for now and come back to this one.